text/0000-app-mixins.md

```diff
@@ -0,0 +1,214 @@
+* If the mixin name is prefixed with `run:` (as per [RFC-0006](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)) the mixin will be dynamically added to the run image.
+* If the mixin name does not have a prefix it will be dynamically added to both the build and run stack images.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When a list of mixins are defined in a `project.toml` and `pack build` is run against that project, the following will happen:
+
+1. The lifecycle will compare the list of mixins to those provided by the stack. If all mixin names are provided by the stack, no further action is required.
+1. In any requested mixin is not provided by the stack, the lifecycle will run the detect phase for all stackpacks defined in the builder.
+1. The lifecycle will compare the mixins added to the build plan to see if they match they required mixins.
+1. If at least one stackpack passes detect and provides the required mixin(s), the lifecycle will execute the stackpack build phase for passing stackpack(s). If no stackpacks pass detect, or no stackpacks provide the required mixin, the build will fail with a message that describes the mixins that could not be provided.
+1. During the lifecycle's build phase, the stackpacks that passed detection will run against the build and run images accordingly (see details below). All stackpacks will be run before the user's buildpacks.
+
+## Stackpacks
+
+ A stackpacks is a special case of buildpack that has the following properties:
+
+* Is run as the `root` user
+* Configured with `privileged = true` in the `buildpack.toml`
+* Can only create one layer
+* Does not have access to the application source code
+* Is run before all regular buildpacks
+* Is run against both the build and run images
+* Is distributed with the builder image
+* May not write to the `/layers` directory
+
+The stackpack interface is identical to the buildpack interface (i.e. the same `bin/detect` and `bin/build` scripts are required). However, some of the context it is run in is different from regular buildpacks.
+
+For each stackpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the stackpack's build phase (excluding `/tmp`, `/cnb`, and `/layers`). Alternatively, a platform may store a new stack image to cache the changes. All of the captured changes will be included in a single layer produced as output from the stackpack. The `/layers` dir MAY NOT be used to create arbitrary layers.
+
+A stackpack is included in a builder by defining it in the `builder.toml` in the `[[stack.buildpacks]]` array of tables:
+
+```
+[[stack.buildpacks]]
+id = "<stackpack id>"
+uri = "<uri to stackpack>"
+```
```

Comment:
I feel like this has been a persistent point of confusion for me. In working group I thought the consensus was that `stackpacks` would ship on the stack images (build and run images) rather than on the builder image.

This would allow us to avoid having to output two images from `create-builder`. We could exclude the stackpacks from the final app image either by adding metadata to the run-image that will allow us to selectively exclude these layers or by adding a layer that [whitesout](https://github.com/opencontainers/image-spec/blob/master/layer.md#whiteouts) the buildpack dir

Reply:
Yea, that's right. I'm wrong in the text of the RFC. Will update

Reply:
@ekcasey do you think the stack should contain the buildpack before the builder image is created, or can the `create-builder` command inject the stackpack(s) (outputting new build and run images w/ the stackpack(s)).

Reply:
I think the build and run images should ship with the stackpacks pre-installed.

Reply:
@ekcasey should i change `uri` to `path`?

Reply:
@jkutner Do we even need to list `stack.buildpacks` if they are already in the image? The `[[buildpacks]]` section typically describe where to fetch buildpacks that should be added to the image.

Is the point to enable/disable them (I wonder if default to enabled is the right call)? In any case I think we only need id/version.

Reply:
I am imagining that stackpacks live either in `/cnb/buildpacks/<id>/<version>` or in an analagous `/cnb/stackpacks`, and can therefore be looked up by `id` and `version`.

Reply:
Addressed in 1881dd3dead0a35b3335c99cfad893d1e2ba24b1